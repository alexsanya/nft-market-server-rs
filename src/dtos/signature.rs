
use serde::{Serialize, Deserialize};

use crate::{models::signature::Signature, utils::patterns::Patterns};

#[derive(Debug, Serialize, Deserialize)]
pub struct SignatureDTO {
    pub v: u64,
    pub r: String,
    pub s: String
}

#[derive(Debug, Clone)]
pub enum ParsingError {
    RisInvalid,
    SisInvalid
}

impl TryInto<Signature> for SignatureDTO {
    type Error = ParsingError;

    fn try_into(self) -> Result<Signature, Self::Error> {
        let patterns = Patterns::new();
        Ok(
            Signature{
                v: self.v,
                r: patterns.test_bytes32(&self.r).map_err(|_| ParsingError::RisInvalid)?.to_owned(),
                s: patterns.test_bytes32(&self.s).map_err(|_| ParsingError::SisInvalid)?.to_owned()
            }
        )
    }
}