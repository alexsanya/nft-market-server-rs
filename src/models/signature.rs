
use ethers::contract::Eip712;
use serde::{Serialize, Deserialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Signature {
    pub v: u64,
    pub r: String,
    pub s: String
}