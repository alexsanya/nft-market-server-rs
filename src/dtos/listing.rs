use serde::{Serialize, Deserialize};
use crate::{models::listing::Listing, utils::patterns::Patterns};
use super::signature::{self, SignatureDTO};

#[derive(Debug, Clone, strum_macros::AsRefStr)]
pub enum ParsingError {
    Owner,
    ChainId,
    NftContract,
    MinPrice,
    TokenId,
    Nonce,
    Signature(signature::ParsingError)
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ListingDTO {
    pub owner: String,
    pub chain_id: String,
    pub min_price_cents: String,
    pub nft_contract: String,
    pub token_id: String,
    pub nonce: String,
    pub signature: SignatureDTO
}

impl TryInto<Listing> for ListingDTO {
    type Error = ParsingError;

    fn try_into(self) -> Result<Listing, Self::Error> {
        let patterns = Patterns::new();
        let listing = Listing{
            owner: patterns.test_address(&self.owner).map_err(|_| ParsingError::Owner)?.to_owned(),
            chain_id: self.chain_id.parse().map_err(|_| ParsingError::ChainId)?,
            min_price_cents: self.min_price_cents.parse().map_err(|_| ParsingError::MinPrice)?,
            nft_contract: patterns.test_address(&self.nft_contract).map_err(|_| ParsingError::NftContract)?.to_owned(),
            token_id: self.token_id.parse().map_err(|_| ParsingError::TokenId)?,
            nonce: self.nonce.parse().map_err(|_| ParsingError::Nonce)?,
            signature: self.signature.try_into().map_err(ParsingError::Signature)?
        };
        Ok(listing)
    }
}

#[cfg(test)]
mod test {
    use insta::assert_debug_snapshot;

    use crate::models::samples::test::get_raw_listing;

    use super::*;

    #[test]
    fn convert() {
        let raw_listing = get_raw_listing();
        let listing_dto: ListingDTO = serde_json::from_value(raw_listing).expect("Cannot convert json to ListingDTO");
        let listing: Listing = listing_dto.try_into().expect("Cannot convert ListingDTO to listing");
        assert_debug_snapshot!(listing);
    }
}