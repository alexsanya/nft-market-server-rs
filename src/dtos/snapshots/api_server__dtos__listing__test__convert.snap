---
source: src/dtos/listing.rs
expression: listing
snapshot_kind: text
---
Listing {
    owner: "0x3897326cEda92B3da2c27a224D6fDCFefCaCf57A",
    chain_id: 11155111,
    min_price_cents: 150000,
    nft_contract: "0xf44b599a0aB6b8cb14E992994BEC0dc59dF883B2",
    token_id: 1,
    nonce: 0,
    signature: Signature {
        v: 28,
        r: "0x5ef4620f4b296763ff15209456d75e868f149a8d1c6821f1ff11fab70bca0ee0",
        s: "0x337ddcb26ea919a2bf5ad6e1d49bd6951a27d1d2e940a5543a70eabc5dbe237e",
    },
}
